### PR TITLE
Fix custom dbcontexts extention methods

### LINKED ISSUE #14893: Entity Framework migrations broken from v12.2 (No database provider has been configured for this DbContext.)
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

12.2

### Bug summary

Upon implementing Entity Framework Core on Umbraco 12.2 as per the documentation (https://docs.umbraco.com/umbraco-cms/tutorials/getting-started-with-entity-framework-core) I'm receiving a DbContext error upon running a first migration. I've attempted the same thing on Umbraco 12.1.2 without fail. 

### Specifics

_No response_

### Steps to reproduce

1. Install Umbraco 12.2.
2. Follow the steps at https://docs.umbraco.com/umbraco-cms/tutorials/getting-started-with-entity-framework-core
3. Not required, but I've set the following in `startup.cs` instead:

```
services.AddUmbracoEFCoreContext<BlogContext>(
                _config.GetConnectionString("umbracoDbDSN") ?? "",
                _config.GetConnectionStringProviderName("umbracoDbDSN") ?? "");
```
4. In a CLI, from the solution root run the following:
`dotnet ef migrations add InitialCreate --context BlogContext --project Website -v`

5. Result:

```
Using project '<path_to_solution>\Website\Website.csproj'.
Using startup project '<path_to_solution>\Website\Website.csproj'.
Writing '<path_to_solution>\Website\obj\Website.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=<path_to_user>\AppData\Local\Temp\tmp2D5C.tmp /verbosity:quiet /nologo "<path_to_solution>\Website\Website.csproj"
Writing '<path_to_solution>\Website\obj\Website.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=<path_to_user>\AppData\Local\Temp\tmp3423.tmp /verbosity:quiet /nologo "<path_to_solution>\Website\Website.csproj"
Build started...
dotnet build "<path_to_solution>\Website\Website.csproj" /verbosity:quiet /nologo

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:04.48
Build succeeded.
dotnet exec --depsfile "<path_to_solution>\Website\bin\Debug\net7.0\Website.deps.json" --additionalprobingpath <path_to_user>\.nuget\packages --runtimeconfig "<path_to_solution>\Website\bin\Debug\net7.0\Website.runtimeconfig.json" <path_to_user>\.dotnet\tools\.store\dotnet-ef\7.0.11\dotnet-ef\7.0.11\tools\net6.0\any\tools\netcoreapp2.0\any\ef.dll migrations add InitialCreate --context BlogContext --assembly "<path_to_solution>\Website\bin\Debug\net7.0\Website.dll" --project "<path_to_solution>\Website\Website.csproj" --startup-assembly "<path_to_solution>\Website\bin\Debug\net7.0\Website.dll" --startup-project "<path_to_solution>\Website\Website.csproj" --project-dir "<path_to_solution>\Website\\" --root-namespace Website --language C# --framework net7.0 --nullable --working-dir "<path_to_solution>" --verbose
Using assembly 'Website'.
Using startup assembly 'Website'.
Using application base '<path_to_solution>\Website\bin\Debug\net7.0'.
Using working directory '<path_to_solution>\Website'.
Using root namespace 'Website'.
Using project directory '<path_to_solution>\Website\'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'Website'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Using application service provider from Microsoft.Extensions.Hosting.
Found DbContext 'UmbracoDbContext'.
Found DbContext 'BlogContext'.
Finding DbContext classes in the project...
Using context 'BlogContext'.
System.InvalidOperationException: No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.Initialize(IServiceProvider scopedProvider, DbContextOptions contextOptions, DbContext context)
   at Microsoft.EntityFrameworkCore.DbContext.get_ContextServices()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType, String namespace)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
No database provider has been configured for this DbContext. A provider can be configured by overriding the 'DbContext.OnConfiguring' method or by using 'AddDbContext' on the application service provider. If 'AddDbContext' is used, then also ensure that your DbContext type accepts a DbContextOptions<TContext> object in its constructor and passes it to the base constructor for DbContext.
```

### Expected result / actual result

With the same code and same installation steps using Umbraco 12.1.2, we have success:

```
dotnet ef migrations add InitialCreate --context BlogContext --project Website -v
Using project '<path_to_solution>\Code\Website\Website.csproj'.
Using startup project '<path_to_solution>\Code\Website\Website.csproj'.
Writing '<path_to_solution>\Code\Website\obj\Website.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=<path_to_user>\AppData\Local\Temp\tmp9264.tmp /verbosity:quiet /nologo "<path_to_solution>\Code\Website\Website.csproj"
Writing '<path_to_solution>\Code\Website\obj\Website.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=<path_to_user>\AppData\Local\Temp\tmp9822.tmp /verbosity:quiet /nologo "<path_to_solution>\Code\Website\Website.csproj"
Build started...
dotnet build "<path_to_solution>\Code\Website\Website.csproj" /verbosity:quiet /nologo
<path_to_solution>\Code\Website\RunDBMigrations.cs(32,17): warning CS4014: Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call. [<path_to_solution>\Code\Website\Website.csproj]

Build succeeded.

<path_to_solution>\Code\Website\RunDBMigrations.cs(32,17): warning CS4014: Because this call is not awaited, execution of the current method continues before the call is completed. Consider applying the 'await' operator to the result of the call. [<path_to_solution>\Code\Website\Website.csproj]
    1 Warning(s)
    0 Error(s)

Time Elapsed 00:00:05.21
Build succeeded.
dotnet exec --depsfile "<path_to_solution>\Code\Website\bin\Debug\net7.0\Website.deps.json" --additionalprobingpath <path_to_user>\.nuget\packages --runtimeconfig "<path_to_solution>\Code\Website\bin\Debug\net7.0\Website.runtimeconfig.json" <path_to_user>\.dotnet\tools\.store\dotnet-ef\7.0.11\dotnet-ef\7.0.11\tools\net6.0\any\tools\netcoreapp2.0\any\ef.dll migrations add InitialCreate --context BlogContext --assembly "<path_to_solution>\Code\Website\bin\Debug\net7.0\Website.dll" --project "<path_to_solution>\Code\Website\Website.csproj" --startup-assembly "<path_to_solution>\Code\Website\bin\Debug\net7.0\Website.dll" --startup-project "<path_to_solution>\Code\Website\Website.csproj" --project-dir "<path_to_solution>\Code\Website\\" --root-namespace Website --language C# --framework net7.0 --nullable --working-dir "<path_to_solution>\Code" --verbose
Using assembly 'Website'.
Using startup assembly 'Website'.
Using application base '<path_to_solution>\Code\Website\bin\Debug\net7.0'.
Using working directory '<path_to_solution>\Code\Website'.
Using root namespace 'Website'.
Using project directory '<path_to_solution>\Code\Website\'.
Remaining arguments: .
Finding DbContext classes...
Finding IDesignTimeDbContextFactory implementations...
Finding application service provider in assembly 'Website'...
Finding Microsoft.Extensions.Hosting service provider...
Using environment 'Development'.
Using application service provider from Microsoft.Extensions.Hosting.
Found DbContext 'UmbracoDbContext'.
Found DbContext 'BlogContext'.
Finding DbContext classes in the project...
Using context 'BlogContext'.
Finding design-time services referenced by assembly 'Website'...
Finding design-time services referenced by assembly 'Website'...
No referenced design-time services were found.
Finding design-time services for provider 'Microsoft.EntityFrameworkCore.SqlServer'...
Using design-time services from provider 'Microsoft.EntityFrameworkCore.SqlServer'.
Finding IDesignTimeServices implementations in assembly 'Website'...
No design-time services were found.
Writing migration to '<path_to_solution>\Code\Website\Migrations\20231002191659_InitialCreate.cs'.
Writing model snapshot to '<path_to_solution>\Code\Website\Migrations\BlogContextModelSnapshot.cs'.
Done. To undo this action, use 'ef migrations remove'
```

### PR DESCRIPTION
**Pull request by nikcio**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes #14893

#### Why

This problem described in #14893 stems from the database provider not being registered when adding a custom DbContext. The reason the problem was introduced in #14674 was because it was unclear what the extension methods usage was. I.E you couldn't easily see that you were supposed to use one method internally (In Umbraco source code) and one from a consuming application.

Therefore, I think that the extension methods needs to be unified and follow the same basic behavioral patterns as the original `AddDbContext()` methods from EF Core. In this way I also think it would be easier for experienced EF Core developers to start using it in an Umbraco context.

To solve this problem this PR adds two new extensions:

```csharp
AddUmbracoDbContext<T>(this IServiceCollection services, Action<DbContextOptionsBuilder>? optionsAction = null)
```

Umbraco equivalent to EF Cores `AddDbContext(this IServiceCollection, Action<DbContextOptionsBuilder>? optionsAction = null, ServiceLifetime contextLifetime = ServiceLifetime.Scoped, ServiceLifetime optionsLifetime = ServiceLifetime.Scoped)`

```csharp
AddUmbracoDbContext<T>(this IServiceCollection services, Action<IServiceProvider, DbContextOptionsBuilder>? optionsAction = null)
```

Umbraco equivalent to EF Cores `AddDbContext(this IServiceCollection, Action<IServiceProvider, DbContextOptionsBuilder>? optionsAction = null, ServiceLifetime contextLifetime = ServiceLifetime.Scoped, ServiceLifetime optionsLifetime = ServiceLifetime.Scoped)`

**Note:**
- Service lifetimes isn't relevant when using `AddUmbracoDbContext` because we use pooled DbContexts after `RuntimeLevel.Run` and transient for every DI request before that.
- `AddUmbracoDbcontext` better matches the EF Core naming of the extension methods.
- Because we don't use a custom `Action` it's easier to tell what parameters do what. And if you have used EF Core before you will be able to use similar setup in the options.

#### What is the behavior of `AddUmbracoDbcontext()`?

Similarly to the normal `AddDbContext` from EF Core this shouldn't magically add the database provider but you should specify what provider to use. To make this behavior similar to what we had in 12.1 I've added two extension methods that can be used in the `optionsAction`.

**Examples:**

```csharp
builder.Services.AddUmbracoDbContext<CustomDbContext>((serviceProvider, options) =>
{
    options.UseUmbracoDatabaseProvider(serviceProvider);
});
```

Uses the Umbraco database connection string

```csharp
builder.Services.AddUmbracoDbContext<CustomDbContext>((options) =>
{
    options.UseDatabaseProvider("My provider name", "Connection string");
});
```

Uses the provider name and connection string provided and automatically adds the correct provider based on the provider name.

```csharp
builder.Services.AddUmbracoDbContext<CustomDbContext>((serviceProvider, options) =>
{
    options.UseSqlite("Data Source=:memory:;Version=3;New=True;");
});
```

Uses the custom provided provider like `AddDbContext` would. This would also make it possible to add options database specific options to the provider.

#### What changed

- This PR includes commit `170ed3a` by Bjarke Berg (The first one) which is cherry picked from `v12/dev`. (This adds the UmbracoPooledDbContextFactory)
- Obsolete the old extension methods
- Added some integration tests to test the registration
- Added new Extension methods to replace the old ones
  - `ÀddUmbracoDbcontext`
  - `UseDatabaseProvider`
  - `UseUmbracoDatabaseProvider`
- Added constant values for the database providers. 

#### What doesn't the PR solve

Right now this PR doesn't auto register the database provider on the extension methods on the old `AddUmbracoEFCoreContext` as this would actually be a breaking change from v12.2 to a version with this change.

But it is quite easy to add we just need to call `UseDatabaseProvider` on the options in the `AddUmbracoEFCoreContext` extension methods.

I'm a bit unsure what the best approach is here.

/CC @Zeegaan

#### Final notes

I think streamlining these extension methods will make it easier to work with EF Core in Umbraco and make a better fit for Umbraco.

Principles of the extension methods:
- A developer should manually specify their database provider like when normally using EF Core but we can provide short hands for using the Umbraco connection string and auto selecting a provider.
- We should support the DbContext itself setting the provider like on the `UmbracoDbContext`.
- We should allow the developer to specify their own database provider so they can parse what ever options they need and they can use the right database for the them.
- The naming should match EF Core so developers don't need to learn new naming
  - And Options should be set in a similarly to EF Core.

I hope it makes sense

#### Test

To test this PR I've added some Unit tests. We could also test other combinations of setting the options in similarly.

### CODE REVIEW

**Comment by emmagarland** [comment]
Hi @nikcio 

Thanks for your PR to fix #14893, where EF Migrations are broken from v12, and for the added tests too!

One of the Core Collaborators team will review this as soon as possible - it also seems like one that we want HQ to review, so I'll contact them about it.

Best wishes

Emma

File: tests/Umbraco.Tests.Integration/Umbraco.Persistence.EFCore/DbContext/CustomDbContextTests.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        {
+        }
+    }
+}
+
+
+[TestFixture]
+[Timeout(60000)]
+[UmbracoTest(Database = UmbracoTestOptions.Database.NewSchemaPerTest, Logger = UmbracoTestOptions.Logger.Console)]
+public class CustomDbContextCustomSqliteProviderTests : UmbracoIntegrationTest
```
Comment (bergmania) [inline tests/Umbraco.Tests.Integration/Umbraco.Persistence.EFCore/DbContext/CustomDbContextTests.cs:50]: Should be marked as `[LongRunning]` if 60000 is required.

File: src/Umbraco.Cms.Persistence.EFCore/Factories/UmbracoPooledDbContextFactory.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.Services;
+
+namespace Umbraco.Cms.Persistence.EFCore.Factories;
+
+/// <inheritdoc/>
+internal class UmbracoPooledDbContextFactory<TContext> : PooledDbContextFactory<TContext>
```
Comment (bergmania) [inline src/Umbraco.Cms.Persistence.EFCore/Factories/UmbracoPooledDbContextFactory.cs:9]: How can this be a new file. It already exists in v12/dev

File: tests/Umbraco.Tests.Integration/Testing/UmbracoIntegrationTest.cs [hunk 3]
```diff
@@ -188,6 +190,13 @@ protected virtual void ConfigureTestServices(IServiceCollection services)
     {
     }
 
+    /// <summary>
+    ///     Hook for registering umbraco test doubles.
+    /// </summary>
+    protected virtual void ConfigureUmbracoTestServices(IUmbracoBuilder builder)
+    {
+    }
```
Comment (bergmania) [inline tests/Umbraco.Tests.Integration/Testing/UmbracoIntegrationTest.cs:198]: Why is this required? Seems to be the same as `CustomTestSetup` called below

**Review by bergmania — CHANGES_REQUESTED** [review]
Added a few comments :)

File: src/Umbraco.Cms.Persistence.EFCore/Factories/UmbracoPooledDbContextFactory.cs (hunk 2 above)
Comment (nikcio) [reply src/Umbraco.Cms.Persistence.EFCore/Factories/UmbracoPooledDbContextFactory.cs:9]: It is the same as the one on `v12/dev` cherry picked from your commit but this PR targets `contrib` where the commit hasn't been merged yet. I cherry picked it to make sure it worked with the Factory.

File: tests/Umbraco.Tests.Integration/Umbraco.Persistence.EFCore/DbContext/CustomDbContextTests.cs (hunk 1 above)
Comment (nikcio) [reply tests/Umbraco.Tests.Integration/Umbraco.Persistence.EFCore/DbContext/CustomDbContextTests.cs:43]: This came from the place I copied the setup for a integration test 😅 
I just tested and it shouldn't be needed so I've removed.

File: tests/Umbraco.Tests.Integration/Testing/UmbracoIntegrationTest.cs (hunk 3 above)
Comment (nikcio) [reply tests/Umbraco.Tests.Integration/Testing/UmbracoIntegrationTest.cs]: I didn't see that when I set up the test. I've switched to `CustomTestSetup` 😉

**Comment by nikcio** [comment]
PR updated 😄 
But what should we do about what I described in "What doesn't the PR solve"?

**Comment by nikcio** [comment]
@bergmania let me know if you're waiting for me to do something else on this PR ☺️

**Comment by Zeegaan** [comment]
@nikcio I think about what the "PR doesn't solve" is okay, as it is obsoleted,
We could even put some info in the obsolete method about it not working properly to warn 🤷 
If you could fix the merge conflict I will get this merged 🚀

**Comment by nikcio** [comment]
@Zeegaan I've now rebased the PR 😄

**Comment by Zeegaan** [comment]
Thanks! I will start testing 🚀

**Comment by Zeegaan** [comment]
Looks good, tests good 🚀

File: src/Umbraco.Cms.Persistence.EFCore/Constants-ProviderNames.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+public static partial class Constants
+{
+    public static class ProviderNames
+    {
+        public const string SQLLite = "Microsoft.Data.Sqlite";
+
+        public const string SQLServer = "Microsoft.Data.SqlClient";
+    }
+}
```
Comment (bjarnef) [inline src/Umbraco.Cms.Persistence.EFCore/Constants-ProviderNames.cs:11]: @Zeegaan would it be better to move the constant to "Umbraco.Core" project, where we have most constants?
https://github.com/nikcio/Umbraco-CMS/tree/1fa09f37fc310c1b14bf68aac2394d592631c8c9/src/Umbraco.Core

Then it can be used without reference to `Umbraco.Cms.Persistence.EFCore`.

Comment (Zeegaan) [reply src/Umbraco.Cms.Persistence.EFCore/Constants-ProviderNames.cs:11]: It can yes, good point 👍

**Comment by jemayn** [comment]
I've found this after starting a new site on 12.3 using EF Core and finding the obsolete mention on the old `AddUmbracoEFCoreContext` method I've used on other sites.

I have looked at the new extension method and I really like the `UseUmbracoDatabaseProvider` method compared to the complexity of the old methods.

However, I have a specific use case that I could solve before that I can't spot how to do with the new extension methods:

If I am running a site where I keep migrations in a seperate project to the Umbraco website project, I could before generate the migrations in that project and register the DI there as long as the project had `Umbraco.Cms.Persistence.EFCore` installed.

However, to make the migrations run I had to specify that the migrations lived in a seperate assembly to the website like this:

In the project with the migrations:

```csharp
public static class DependencyInjection
{
    public static IServiceCollection AddPersistence(this IServiceCollection services)
    {
        services.AddUmbracoEFCoreContext<FavouriteListContext>((options, connectionString, providerName) => options.UseSqlServer(connectionString,
            builder =>
            {
                builder.MigrationsAssembly(typeof(DependencyInjection).Assembly.GetName().FullName);
            }));

        return services;
    } 
}
```

And then in the websites startup file:

```csharp
services.AddUmbraco(_env, _config)
            .AddBackOffice()
            .AddWebsite()
            .AddComposers()
            .Build();
        
services.AddPersistence();
```

Now with the new methods I can get the connection string for the Umbraco db automatically but not set the assembly:

```csharp
services.AddUmbracoDbContext<MovieContext>((serviceProvider, options) =>
{
    options.UseUmbracoDatabaseProvider(serviceProvider);
});
```

Is there a way to both set the database to be the same, and also still configure the assembly with the new extension methods?

**Comment by nikcio** [comment]
@jemayn For a quick fix that closely matches what you had with the previous extension method you could use:
```csharp
services.AddUmbracoDbContext<MovieContext>((serviceProvider, options) =>
{
    ConnectionStrings connectionStrings = serviceProvider.GetRequiredService<IOptionsMonitor<ConnectionStrings>>().CurrentValue;

    // Replace data directory
    string? dataDirectory = AppDomain.CurrentDomain.GetData(Constants.System.DataDirectoryName)?.ToString();
    if (string.IsNullOrEmpty(dataDirectory) is false)
    {
        connectionStrings.ConnectionString = connectionStrings.ConnectionString?.Replace(Constants.System.DataDirectoryPlaceholder, dataDirectory);
    }

    if (string.IsNullOrEmpty(connectionStrings.ProviderName))
    {
        Log.Warning("No database provider was set. ProviderName is null");
        return;
    }

    if (string.IsNullOrEmpty(connectionStrings.ConnectionString))
    {
        Log.Warning("No database provider was set. Connection string is null");
        return;
    }

    builder.UseSqlServer(connectionString, x => 
    {
        x.MigrationsAssembly(typeof(MovieContext).Assembly.FullName);
    });
});
```

We could make this easier by taking extracting the getting the Umbraco connection string to it's own extension method so that `UseUmbracoDatabaseProvider` would just call `builder.GetUmbracoDatabaseConnection` (or whatever naming makes sense) and `builder.UseDatabaseProvider`.

**Comment by nikcio** [comment]
@jemayn I've opened a PR here: https://github.com/umbraco/Umbraco-CMS/pull/15186 to make the above example easier :smile:
